07/AcyclicGraph.scala

```diff
@@ -0,0 +1,214 @@
+      if (requirements.isEmpty) steps
+      else {
+        val (ind, dep): (Step, Step) = parseStep(requirements.head)
+        val intermediate: List[Step] = updateInPlace(steps, ind)
+        go(requirements.tail, updateInPlace(intermediate, dep))
+      }
+    }
+    go(requirements, List())
+  }
+
+  def updateSatisfiedDependencies(steps: List[Step], completed: Step): List[Step] = {
+    def checkStep(step: Step): Step = {
+      def go(dependencies: List[Dependency], updated: List[Dependency]): Step = {
+        if (dependencies.isEmpty) Step(step.name, updated)
+        else if (dependencies.head.name == completed.name) go(dependencies.tail, Dependency(completed.name, true) :: updated)
+        else go(dependencies.tail, dependencies.head :: updated)
+      }
+      go(step.dependencies, List())
+    }
+    def go(steps: List[Step], updated: List[Step]): List[Step] = {
+      if (steps.isEmpty) updated
+      else go(steps.tail, checkStep(steps.head) :: updated)
+    }
+    go(steps, List())
+  }
+
+  def updateSatisfiedDependenciesList(steps: List[Step], completed: List[Step]): List[Step] = {
+    if (completed.isEmpty) steps
+    else updateSatisfiedDependenciesList(updateSatisfiedDependencies(steps, completed.head), completed.tail)
+  }
+
+  def clearExecuted(steps: List[Step], completed: Step): List[Step] = {
+    def go(steps: List[Step], updated: List[Step]): List[Step] = {
+      if (steps.isEmpty) updated
+      else go(steps.tail, if (steps.head.name == completed.name) updated else steps.head :: updated)
+    }
+    go(steps, List())
+  }
+
+  def clearExecutedList(steps: List[Step], completed: List[Step]): List[Step] = {
+    if (completed.isEmpty) steps
+    else clearExecutedList(clearExecuted(steps, completed.head), completed.tail)
+  }
+
+  def executeNext(steps: List[Step]): Step = {
+    def go(steps: List[Step], readySteps: List[Step]): Step = {
+      if (steps.isEmpty) {
+        if (readySteps.size == 1) readySteps(0)
+        else readySteps.sortWith(_.name < _.name).head
+      } else go(steps.tail, if (steps.head.isReady()) steps.head :: readySteps else readySteps)
+    }
+    go(steps, List())
+  }
+
+  def getAllExecutableSteps(steps: List[Step], workers: List[Worker]): List[Step] = {
+    def checkWorkers(w: List[Worker], s: Step): Boolean = {
+      if (w.isEmpty) false
+      else if (w.head.step.name == s.name) true
+      else checkWorkers(w.tail, s)
+    }
+
+    def go(steps: List[Step], readySteps: List[Step]): List[Step] = {
+      if (steps.isEmpty) readySteps.sortWith(_.name < _.name)
+      else if (steps.head.isReady()) {
+        if (checkWorkers(workers, steps.head)) go(steps.tail, readySteps)
+        else go(steps.tail, steps.head :: readySteps)
+      } else {
+        go(steps.tail, readySteps)
+      }
+    }
+    go(steps, List())
+  }
+
+  def performExecution(steps: List[Step]): String = {
+    def go(steps: List[Step], acc: String): String = {
```

Comment:
Strings are generally not good as accumulators. You might prefer using `List[Char]` and then convert to `String` (`_.reverse.mkString`).